scikeras/wrappers.py

```diff
@@ -432,10 +499,38 @@ def preprocess_X(X):
             X : unchanged 2D numpy array
             extra_args : attributes of output `y`.
         """
-        extra_args = dict()
+        extra_args = {
+            "X_dtype_": X.dtype,
+            "X_shape_": X.shape,
+        }
         return X, extra_args
 
-    def fit(self, X, y, sample_weight=None, warm_start=False, **kwargs):
+    def fit(self, X, y, sample_weight=None):
+        """Constructs a new model with `build_fn` & fit the model to `(X, y)`.
+
+        Arguments:
+            X : array-like, shape `(n_samples, n_features)`
+                Training samples where `n_samples` is the number of samples
+                and `n_features` is the number of features.
+            y : array-like, shape `(n_samples,)` or `(n_samples, n_outputs)`
+                True labels for `X`.
+            sample_weight : array-like of shape (n_samples,), default=None
+                Sample weights. The Keras Model must support this.
+        Returns:
+            self : object
+                a reference to the instance that can be chain called
+                (ex: instance.fit(X,y).transform(X) )
+        Raises:
+            ValueError : In case of invalid shape for `y` argument.
+        """
+        return self._fit(
+            X=X,
+            y=y,
+            sample_weight=sample_weight,
+            warm_start=getattr(self, "warm_start", False),
```

Comment:
This would break subclassed models that do not have the `warm_start` param

Reply:
I don't think SciKeras should protect against violations of the [Liskov substitution principle][1], which says the following:

> Objects in a program should be replaceable with instances of their subtypes without altering the correctness of that program.
>
> —[SOLID][2]

I think SciKeras can safely assume that all models are children of `BaseWrapper`. All that's required for that is calling `super().__init__` in `Child.__init__`.

[1]:https://en.wikipedia.org/wiki/Liskov_substitution_principle
[2]:https://en.wikipedia.org/wiki/SOLID

Reply:
I felt that if we don't _require_ the `warm_start` parameter the fact that the child class does not use it is not a violation. My intention wasn't to protect against a violation, but rather to tell the users that they can safely subclass and _not_ call `super.__init__()` without breaking anything or violating LSP. In other words, since you can substitute a `BaseWrapper` for `Child` even if `Child` does not call `super().__init__` then there is not violation.

But my background is not CS, so my understanding of this may be off.

Reply:
`warm_start` is a parameter inside `BaseWrapper.__init__` on the most recent commit in this PR: https://github.com/adriangb/scikeras/blob/c5d9d32091e3f9f3a6414a7f0db2ed05f525ad06/scikeras/wrappers.py#L134

I think if `warm_start` is used in the implementation, it should be a parameter.

> tell the users that they can safely subclass and not call `super.__init__()` without breaking anything or violating LSP

Not calling `super().__init__` is itself a violation of LSP.

LSP says "an instance of the base class and instances of subclasses should all have the same attributes as the base class" so they can be treated equivalently. For Scikit-Learn, `super().__init__` is where all the parameters get set. For example, this code fails:

``` python
from sklearn.linear_model import SGDClassifier
from sklearn.datasets import make_classification

class Custom(SGDClassifier):
    def __init__(self, foo=3):
        self.foo = foo

est = Custom()
X, y = make_classification()
est.fit(X, y)  # attribute error: 'Custom' object has no attribute 'alpha'
```

Additionally, all the examples in the documentation should call `super().__init__`. It's best practice.

Reply:
I am not sure where that quote came from. I feel that this may fall into forcing an implementation vs. forcing an interface. In my opinion, all of the parameters that we use in `BaseWrapper.__init__` are optional (they have a default value) and so excluding them does not and should not break the interface.

This said, I do believe that enforcing cooperative inheritance aligns much nicer with other languages, since Python is the exception to the rule in terms of not forcing implicit calling of the parent class' constructor (referring to __new__ + __init__ here).

It is also true that if we enforce this now and remove this later, nothing will break. On the other hand, if we do not enforce this now and need it later (for reasons I cannot think of now but also cannot rule out) it might break things for everyone to add it. This is a practical argument, which for me is much more important than an abstract one. Because of this, I decided to make the `getattr -> self.` replacement and enforce calling of the parent's constructor.

I checked the docs, I believe the current docs are using `super.__init__()` everywhere. I found one test that was not and fixed it. I will keep this in mind when I update #73

Reply:
One practical of this is that all child classes must either pass `**kwargs` or explicitly put every single parameter from `BaseWrapper` into their constructor. The following is not valid:

```python3
class SubclassedClassifier(KerasClassifier):
    def __init__(self, param1=None):
        super().__init__()
        self.param1 = param1
```

Instead users would have to use:

```python3
class SubclassedClassifier(KerasClassifier):
    def __init__(self, param1=None, **kwargs):
        super().__init__(**kwargs)
        self.param1 = param1
```

Which is okay, but I do think being able to do:

```python3
class SubclassedClassifier(KerasClassifier):
    def __init__(self, param1=None):
        self.param1 = param1
```

Results in clearer code and would be easier for to hint and such.